ui/desktop/src/hooks/useChatStream.ts

```diff
@@ -145,47 +209,60 @@ export function useChatStream({
   sessionId,
   onStreamFinish,
 }: UseChatStreamProps): UseChatStreamReturn {
-  const [messages, setMessages] = useState<Message[]>([]);
+  const [messages, _setMessages] = useState<Message[]>([]);
   const messagesRef = useRef<Message[]>([]);
   const [session, setSession] = useState<Session>();
   const [sessionLoadError, setSessionLoadError] = useState<string>();
   const [chatState, setChatState] = useState<ChatState>(ChatState.Idle);
   const abortControllerRef = useRef<AbortController | null>(null);
-
-  const setMessagesAndLog = useCallback(
-    (messages: Message[], log: string) => {
-      console.log(log, session, messages.length);
-      setMessages(messages);
-    },
-    [session]
-  );
-
-  useEffect(() => {
-    if (session) {
-      resultsCache.set(sessionId, { session, messages });
-    }
-  }, [sessionId, session, messages]);
+  const sessionIdRef = useRef<string>(sessionId);
 
   const renderCountRef = useRef(0);
   renderCountRef.current += 1;
   console.log(`useChatStream render #${renderCountRef.current}, ${session?.id}`);
 
+  const setMessagesAndLog = useCallback((newMessages: Message[], logContext: string) => {
+    log.messages(logContext, newMessages.length, {
+      lastMessageRole: newMessages[newMessages.length - 1]?.role,
+      lastMessageId: newMessages[newMessages.length - 1]?.id?.slice(0, 8),
+    });
+    _setMessages(newMessages);
+    messagesRef.current = newMessages;
+  }, []);
+
   useEffect(() => {
-    messagesRef.current = messages;
-  }, [messages]);
+    if (sessionIdRef.current !== sessionId) {
+      log.session('changed', sessionId, {
+        from: sessionIdRef.current.slice(0, 8),
+        to: sessionId.slice(0, 8),
+      });
+      sessionIdRef.current = sessionId;
+      setMessagesAndLog([], 'session-reset');
+      setSession(undefined);
+      setSessionLoadError(undefined);
+      setChatState(ChatState.Idle);
+    }
+  }, [sessionId, setMessagesAndLog]);
```

Comment:
this useEffect has the same dependencies as the one below. I don't think there's a reason to have both. it also introduces a sessionIdRef that only checks if the sessionId has changed, but the sessionId is in the dependencies, so the useEffect already only runs when sessionId changes, no need for that ref.

I think all we want to do is add the resets to the block below